### DISCUSSION TITLE
How can I use Cache in List / Show ?

### DISCUSSION THREAD

**Discussion by magic-thomas** [filamentphp/filament#6977] [Help]
Hi. 

I am new to fliament. 

I would like to load data from Cache in List / Show page. or table component. 

In which method can I put Cache functions ?

**Comment by pboivin** [comment]
The Table on the List page needs to work with an Eloquent query, you can't use cached data.
Maybe this article can give you an idea: https://filamentphp.com/blog/how-to-consume-an-external-api-with-filament-tables

**Reply by magic-thomas** [reply]
That is too sad to hear. 

That might mean Filament package is not proper to CMS or Frontend page. 

I wish I can find hacky way .

Thanks.

**Reply by pboivin** [reply]
It would help to know what you are actually trying to do... There might be a solution to your problem that is not a hack. Also read the article I shared, there is a clean way to use the full Table Builder with static data.

**Reply by danharrin** [reply]
How is it "not a proper CMS"? What is taking so long that you need to cache?

**Reply by magic-thomas** [reply]
When I build web site, usually it has a lot of relationship. 

I will explain with my case .

In index page, there are some models need. Post, comments, like, User, Category, Menu ( for making url ), Image, Point, bookmark, ClickLog, Like, User, Role, Permission, etc 

Yes,  we might have to avoid using cache for index page, usually.  

But, it is not easy to avoid using cache for show page.

For show page, it need more relationships.

For example, When visitor navigate a show Post Model Page, it should be with Category, Image, Point, Menu, User, User.role, user.permission, UserProfile, Like, Comment, comments.users, comments.likes, ClickLog, Bookmark , etc. 

And what if the comments has hierarchical structure? It will be more complicated.

And there are list of best read posts within a week and a month in that page.

And there is a list of posts of same page number.

If the Cache is not applied, whenever visitor open page, it will be more than 50 queries even though  Eager Loading is applied.

So, I think it is better to use Cache for performance. specially for building CMS.

**Reply by pboivin** [reply]
Filament has no opinion on how you build the rest of your site really... you can absolutely use Cache or ResponseCache on your user-facing pages.

**Comment by MhdTarhini** [comment]
Hello, did you manage to apply caching on the list page?

**Comment by oliverbj** [comment]
Also in dire need of caching...

**Comment by deven-sitapara** [comment]
I have build my app 95% and now I read this thread so sad to head that

**Comment by hansensalim** [comment]
My solution for this is to utilise a custom eloquent Builder

```
namespace App\Models;

use App\Database\CustomBuilder;
use Illuminate\Database\Eloquent\Model;

class MyModel extends Model
{
    public function newEloquentBuilder($query)
    {
        return new CustomBuilder($query);
    }
}
```

```
namespace App\Database;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Pagination\Paginator;

class CustomBuilder extends Builder
{
    protected ?string $cacheKey     = null;
    protected int     $cacheMinutes = 10;

    public function withCache(string $key, ?int $minutes = 10): self
    {
        $this->cacheKey     = $key;
        $this->cacheMinutes = $minutes;
        return $this;
    }

    // This is not the actual caching strategy i use, just some example.
    // Just be conscious with how you will manage cache of paginated data.
    public function paginate($perPage = null, $columns = ['*'], $pageName = 'page', $page = null, $total = null)
    {
        if (!$this->cacheKey) {
            return parent::paginate($perPage, $columns, $pageName, $page, $total);
        }

        $page = $page ?? Paginator::resolveCurrentPage($pageName);
        $paginationHash = md5(json_encode(compact('perPage', 'page', 'columns', 'pageName', 'total')));

        $paginatedCacheKey = $this->cacheKey . $paginationHash;
        $cachedData = cache()->get($paginatedCacheKey, []);

        if (empty($cachedData)) {
            $cachedData = parent::paginate($perPage, $columns, $pageName, $page, $total);
            cache()->put($paginatedCacheKey, $cachedData, now()->addMinutes($this->cacheMinutes));
        }

        return $cachedData;
    }
}
```

And finally on the resource class
```
class MyModelResource extends Resource
{

    public static function table(Table $table): Table
    {
        return $table
            ->modifyQueryUsing(function (CustomBuilder $query) {
                $query->withCache(key: 'my-cache-key-for-this-resource', minutes: 15);
            })
            ->columns([...]);
    }
```
